File: src/main/scala/com/tiny/lang/scala/generic/variance/demo/Box.scala
package com.tiny.lang.scala.generic.variance.demo

import scala.collection.mutable.ListBuffer

/**
  * instances of Box are read-only because of covariance
  *
  */
class Box[+T](items: T*) {

  /**
    * compile error: because of covariance is readonly
    *
    * class Box[+T]
    * val list: ListBuffer[_ <: T] = new ListBuffer[T]
    * // instances of parent can't add to the List of child
    * def put[U >: T](u: U)(op: U => T) = list += op(u)
    * def addList(lis: ListBuffer[_ <: T]): Unit = list ++= lis
    */

  def head: T = items.head

  def foreach[U](f: T => U): Unit = items.foreach(f)

  def accept[R](boxVisitor: BoxVisitor[T, R]): R = {
    val list = new ListBuffer[R]
    foreach((t: T) => list += boxVisitor.visit(t))
    list.head
  }

}

object Box {

  def apply[T](ts: T*): Box[T] = new Box(ts: _*)

}

class Fuel

class Plant extends Fuel

class Bamboo extends Plant

class Energy

File: src/main/java/com/tiny/lang/java/generic/variance/Box.java
package com.tiny.lang.java.generic.variance;

/**
 * 协变
 *
 * @author tiny.wang
 */
public class Box<T> {

    /**
     * 若有一个Box<Plant>，可以用BoxVisitor<Fuel,_>, BoxVisitor<Plant,_>来消费他，
     * 不能用BoxVisitor<Bamboo,_>来消费，即泛型T逆变；
     * <p>
     * 对于返回值：
     * 默认情况下Java的多态规则本身就契合协变规则：子类可当父类使用
     * 所以此处的<_, ? extends R>与<_, R>在当前情况下并无差异
     *
     * @param visitor visitor
     * @param <R>     R
     * @return value
     */
    public <R> R accept(BoxVisitor<? super T, ? extends R> visitor) {
        return null;
    }

    /**
     * entity classes
     */

    static class Fuel {
    }

    static class Plant extends Fuel {
    }

    static class Bamboo extends Plant {

    }
}
